internal/runtime/catalog/openapi.go

```diff
@@ -0,0 +1,249 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package catalog
+
+import (
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// OpenAPI generates and returns the OpenAPI spec.
+func (c *Catalog) OpenAPI(version string) (*spec3.OpenAPI, error) {
+	openAPI := &spec3.OpenAPI{
+		Version: "3.0.0",
+		Info: &spec.Info{
+			InfoProps: spec.InfoProps{
+				Description: "Open API specification for Cluster API Runtime SDK",
+				Title:       "Cluster API - Runtime SDK",
+				License: &spec.License{
+					Name: "Apache 2.0",
+					URL:  "http://www.apache.org/licenses/LICENSE-2.0.html",
+				},
+				Version: version,
+			},
+		},
+		Paths: &spec3.Paths{
+			Paths: map[string]*spec3.Path{},
+		},
+		Components: &spec3.Components{
+			Schemas: map[string]*spec.Schema{},
+		},
+	}
+
+	for gvh, hookDescriptor := range c.gvhToHookDescriptor {
+		err := addHookAndTypesToOpenAPI(openAPI, c, gvh, hookDescriptor)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return openAPI, nil
+}
+
+func addHookAndTypesToOpenAPI(openAPI *spec3.OpenAPI, c *Catalog, gvh GroupVersionHook, hookDescriptor hookDescriptor) error {
+	// Create the operation.
+	operation := &spec3.Operation{
+		OperationProps: spec3.OperationProps{
+			Tags:        hookDescriptor.metadata.Tags,
+			Summary:     hookDescriptor.metadata.Summary,
+			Description: hookDescriptor.metadata.Description,
+			OperationId: operationID(gvh),
+			Responses: &spec3.Responses{
+				ResponsesProps: spec3.ResponsesProps{
+					StatusCodeResponses: make(map[int]*spec3.Response),
+				},
+			},
+			Deprecated: hookDescriptor.metadata.Deprecated,
+		},
+	}
+	path := GVHToPath(gvh, "")
+
+	// Add name parameter to operation path if necessary.
+	// TODO(openapi): TBD if we want to add the name parameter.
+	// I think it's useful as that reflects the real path which Runtime Extensions will handle.
+	// Follow-up: potentially drop Singleton field.
+	if !hookDescriptor.metadata.Singleton {
+		path = GVHToPath(gvh, "{name}")
+		operation.Parameters = append(operation.Parameters, &spec3.Parameter{
+			ParameterProps: spec3.ParameterProps{
+				Name:     "name",
+				In:       "path",
+				Required: true,
+				Schema: &spec.Schema{
+					SchemaProps: spec.SchemaProps{
+						Type: []string{"string"},
+					},
+				},
+			},
+		})
+	}
+
+	// Add request type to operation.
+	requestGVK, err := c.Request(gvh)
+	if err != nil {
+		return err
+	}
+	requestType, ok := c.scheme.AllKnownTypes()[requestGVK]
+	if !ok {
+		return errors.New("TODO 4")
```

Comment:
please go over the error returns and make sure we add enough context when necessary (and please replace my TODO messages :) )